icu4j/main/classes/core/src/com/ibm/icu/impl/units/UnitConverter.java

```diff
@@ -0,0 +1,309 @@
+        Convertibility convertibility = extractConvertibility(source, target, conversionRates);
+        assert (convertibility == Convertibility.CONVERTIBLE || convertibility == Convertibility.RECIPROCAL);
+
+        Factor sourceToBase = conversionRates.getFactorToBase(source);
+        Factor targetToBase = conversionRates.getFactorToBase(target);
+
+        if (convertibility == Convertibility.CONVERTIBLE) {
+            this.conversionRate = sourceToBase.divide(targetToBase).getConversionRate();
+        } else {
+            this.conversionRate = sourceToBase.multiply(targetToBase).getConversionRate();
+        }
+
+        // calculate the offset
+        this.offset = conversionRates.getOffset(source, target, sourceToBase, targetToBase, convertibility);
+    }
+
+    static public Convertibility extractConvertibility(MeasureUnitImpl source, MeasureUnitImpl target, ConversionRates conversionRates) {
+        ArrayList<SingleUnitImpl> sourceSingleUnits = conversionRates.getBasicUnitsWithoutSIPrefix(source);
+        ArrayList<SingleUnitImpl> targetSingleUnits = conversionRates.getBasicUnitsWithoutSIPrefix(target);
+
+        HashMap<String, Integer> dimensionMap = new HashMap<>();
+
+        insertInMap(dimensionMap, sourceSingleUnits, 1);
+        insertInMap(dimensionMap, targetSingleUnits, -1);
+
+        if (areDimensionsZeroes(dimensionMap)) return Convertibility.CONVERTIBLE;
+
+        insertInMap(dimensionMap, targetSingleUnits, 2);
+        if (areDimensionsZeroes(dimensionMap)) return Convertibility.RECIPROCAL;
+
+        return Convertibility.INCONVERTIBLE;
+    }
+
+    /**
+     * Helpers
+     */
+    private static void insertInMap(HashMap<String, Integer> dimensionMap, ArrayList<SingleUnitImpl> singleUnits, int multiplier) {
+        for (SingleUnitImpl singleUnit :
+                singleUnits) {
+            if (dimensionMap.containsKey(singleUnit.getSimpleUnit())) {
+                dimensionMap.put(singleUnit.getSimpleUnit(), dimensionMap.get(singleUnit.getSimpleUnit()) + singleUnit.getDimensionality() * multiplier);
+            } else {
+                dimensionMap.put(singleUnit.getSimpleUnit(), singleUnit.getDimensionality() * multiplier);
+            }
+        }
+    }
+
+    private static boolean areDimensionsZeroes(HashMap<String, Integer> dimensionMap) {
+        for (Integer value :
+                dimensionMap.values()) {
+            if (!value.equals(0)) return false;
+        }
+
+        return true;
+    }
+
+    public BigDecimal convert(BigDecimal inputValue) {
+        return inputValue.multiply(this.conversionRate).add(offset);
+    }
+
+    public enum Convertibility {
+        CONVERTIBLE,
+        RECIPROCAL,
+        INCONVERTIBLE,
+    }
+
+    /**
+     * Responsible for all the Factor operation
+     * NOTE:
+     * This class is immutable
+     */
+    static class Factor {
+        private BigDecimal factorNum;
+        private BigDecimal factorDen;
+        /* FACTOR CONSTANTS */
```

Comment:
This comment makes it sound like these ints should be the factors themselves, e.g. pi being 3.14159... - rather than as the powers. I commented on this in May, but not too clearly, and I reckoned we could leave it for later.

Do you want to fix the documentation now, or do you want to leave it for more work later (i.e. icu-units#84)? If you'd like to leave it for later, how about we add a TODO at the top of this file for improving the documentation?

The point of documentation is to make it easier for future readers of the code to understand. When the documentation only manages to confuse, no documentation is better.

In C++ we have `int32_t constants[CONSTANTS_COUNT] = {};` and we index them via constants.

Here in Java, you're doing this: `result.CONSTANT_GRAVITY = this.CONSTANT_GRAVITY * power;`
So we're redefining gravity as gravity^power. I didn't like "constants" as a variable name in C++, here it's simply worse.

I know finding good names isn't easy. "POWER_GRAVITY" and "POWER_PI" feel weird too, but it's at least closer to what these values mean, the power to which these constant factors are raised, rather than it looking like we're redefining universal constants.

Reply:
This is the best I've come up with so far, and I'd be much more satisfied with it than CONSTANT_*: EXPONENT_GRAVITY, EXPONENT_PI, ...

Reply:
I agree with you, I have some ideas about that, for now, I have just added a TODO, because of the hard deadlines

Reply:
I've created https://github.com/unicode-org/icu/pull/1464 to address this.